### ISSUE TITLE
[Feature Request] Should be possible to init Ghost with a config object

### ISSUE THREAD

**Issue by coen-hyde** [TryGhost/Ghost#4731]
I'm using Ghost as a module and it would be good if I could initialize or update Ghost with my own config object, rather than providing Ghost with a filepath for where my config file is located. I have my own way of storing configs.

**Comment by ErisDS** [comment]
Hi @coen-hyde, welcome to Ghost's GitHub. 

I see you've got both an issue and a PR, and you've closed the issue in favour of the PR but I'm commenting here because the issue is really the best place to discuss the merits of the feature request, and also we usually use GitHub's features to close an issue with a PR if it is decided it's something we want to add. 

We've got some [contributing guidelines](https://github.com/TryGhost/Ghost/blob/master/CONTRIBUTING.md) that might be worth a bit of a read - both in relation to feature requests and the format of your PR, but in short, my main concern here is that neither the issue you've raised not the PR really explain your use case or why it would be worthwhile to add this extra level of complexity to Ghost.

There's already a couple of different ways to specify the path to a config file, including passing in a path or setting an environment variable, and there is no discussion of why these are not suitable. Adding yet-another-way to specify config adds quite significant complexity especially when you consider how rare the use case is  - it's something that's easy to end up getting broken in future. Therefore I'd like to be convinced that there is a definite use case for this.

**Comment by esatterwhite** [comment]
I would have to second this PR. The very first thing I tried to do was require ghost and pass it configuration options collected from the rest of my app. 

It wouldn't be much of an mess to take options at start up. The problem looks to stem from [here](https://github.com/TryGhost/Ghost/blob/master/core/server/config/index.js#L404). The config handler exports a singleton and it will only read from a hard coded path for testing. 

This could be made pretty flexible with out a lot of code with something like [nconf](https://www.npmjs.com/package/nconf) or [confit](https://www.npmjs.com/package/confit). 

This would make multi-tier config pretty straight forward and make ghost as a module really simple to set up and run.

This is how we [use nconf in our production apps](https://gist.github.com/esatterwhite/14e4361d0d49865ce3d2) 

minus all of the documentation in there, this is less than 100 lines of code.

This conf module checks for:
1. System level overrides
2. Command line arguments
3. Environment variables
4. A configuration file(s)
5. System specified defaults

This could easily be extended to key off of NODE_ENV and read config.<ENV>.json, etc if that is your cup of tea. And this would really be a matter of changing `config.foo` to `+config.get('foo')`

Doing something like this would free developers to set up, deploy and configure ghost how ever they wanted. It would also mean that I don't have to keep massive config files with sensitive authentication information in source control.

**Comment by esatterwhite** [comment]
@coen-hyde So In looking into configuration happens, it looks like the complexity that @ErisDS talking about stems from how the configuration is actually handled. Its a Mix of settings and functionality, and that functionality is required to establish the settings. 

Additionally the Database connection is created during the configuration after initialization during an `init` call. Which makes things a bit weird, because you can't require/use any Database models until the config singleton has been loaded and initialized, which creates a new database connection. 

The Way the configuration / database connection is set up is mostly what makes this difficult to use as a module because the order of things really matters despite the large amount of flow control with promises. if you don't `config.load` -> `models.init` -> `migrations.init` -> `settings.populateDefaults` -> `api.init` -> `permissions.init`  You've got problems.



 I think what would make all of the "complexity" significantly easier would be to consider config as a single read only object of settings, and nothing more where the act of requiring it establishes resolves environment variables, overrides, files, command line arguments  and sane defaults etc into a single object.

Then the act of requiring a database model -> requires the database connection -> requires config, which has the correct settings, and the defaults for GhostServer are the result of config resolution, unless over rides are passed

```
server = new GhostServer() == new GhostServer( require( config ) )
server.start()
```

right now it is actually
```
require('config').init()
config.load( /path/to/config )[NODE_ENV]
config.set( overrides )
models.init();
migrations.init();
models.Settings.populateDefaults();
api.init();
permissions.init();
```

And this all stems from the notion that configuration is `settings`, `functionality` and a `database connection` mashed into thing. Additionally, their are effectively no defaults, and you have to set `NODE_ENV` and specify a file path to get something marginally working.

In short, config.js is messy and should be re-thought.

using [nconf](https://www.npmjs.com/package/nconf) and treating config as a single read only object  would make most of this go away.

the config module is reduced to 

```
module.exports = new nconf.Provider()
        .overrides( {...} )
        .argv()
        .env({separator:'__'})
        .file( your/optional/config/file ) // non - existent files are ignored
        .defaults( {... } )
```

db connection is `new knex( conf.get('database') )`

The current set up made a lot of sense when Ghost was stand alone, but as a module, it will need to stop trying to do everything at start up.

**Comment by pensierinmusica** [comment]
+1 https://github.com/TryGhost/Ghost/issues/5145

**Comment by ErisDS** [comment]
Whilst I am in agreement that there needs to be improvement around how we handle config in Ghost, i think @esatterwhite hit an important nail on the head here: 

> if you don't config.load -> models.init -> migrations.init -> settings.populateDefaults -> api.init -> permissions.init You've got problems.

The real, underlying problem is more to do with how config & Ghost itself is initialised / booted up i.e. #2182, and that solving those problems should help us to move towards being able to use something like nconf in a manageable way.

**Comment by iamjochem** [comment]
+1 - ghost as a module should not dictate how a config object is supplied - I'm running ghost as a module and my app config is initialized elsewhere, currently I have load my config (it's promise-based because it does some async data-retrieval of it's own) build a config object - write that into a file (in such a way that is acts as a valid node-module) and specify this file as ghost's config file ... which is crufty.

the fact that ghost needs to run a host of initialization steps (`config.load -> models.init -> migrations.init -> settings.populateDefaults -> api.init -> permissions.init`) has no real technical baring on how a "raw config object" is supplied.

if ghost is given a config object it should use in the same way it normally uses the result of the following:

```js
ConfigManager.prototype.readFile = function (envVal) {
    return require(this._config.paths.config)[envVal];
};
```
(taken from [https://github.com/TryGhost/Ghost/blob/master/core/server/config/index.js#L318]())

I get the impression that this could easily be implementing with a small change in `configManager.load`:

```js
ConfigManager.prototype.load = function (configFilePath) {
    var self = this;
  
    self._config.paths.config = process.env.GHOST_CONFIG || configFilePath || self._config.paths.config;

    /* Check for config file and copy from config.example.js
        if one doesn't exist. After that, start the server. */
    return new Promise(function (resolve, reject) {
/** 
 * CHECK FOR LITERAL OBJECT HERE ...
 */ 
           if (_.isObject(configFilePath))
               resolve(self.init(configFilePath));

        fs.stat(self._config.paths.config, function (err) {
            var exists = (err) ? false : true,
                pendingConfig;

            if (!exists) {
                pendingConfig = self.writeFile();
            }

            Promise.resolve(pndingConfig).then(function () {
                return self.validate();
            }).then(function (rawConfig) {
                resolve(self.init(rawConfig));
            }).catch(reject);
        });
    });
};
```

> **P.S. on an unrelated note** - you may or may not be aware of the fact that the author of bluebird considers the `new Promise(function(resolve, reject ...` strategy to be an anti-pattern: [http://bluebirdjs.com/docs/anti-patterns.html]() ... I personally believe you are better off using `bluebird.promisify()` on `fs.stat` in this case.

**Comment by juanpaco** [comment]
1st, thanks for Ghost!  And for making it free and open source.

I also like the idea of being able to supply a config object rather than requiring Ghost to load the config from a file.

In my case I'm using Ghost as an npm module and mounting it into an existing Express app.  I'm also deploying to Heroku and am using the 12-factor app pattern.  My app's startup sequence reads all env variables in one location and other parts use that to build whatever objects they need.  Having this one location gives me a single place to easily check if important env variables are set.  I'd like to similarly construct a config object for Ghost and just pass it in, then Ghost doesn't have to aware of how the config is stored-- it just gets an object.

@iamjochem 's impression seems like a good strategy to me, albeit I have not worked on the Ghost source at all.  The `load` function seems to do a couple of different things, one of which is identifying where the config settings come from.  It seems how the config object is retrieved is a separate concern from validating said object, so checking to see if an object was provided or if a path was provided would be self-contained.

If this got approved as a feature, I'm quite happy to help with the work on it, as it seems the other folks on this thread are too.

Thanks!

**Comment by ErisDS** [comment]
I've taken on board the discussion in this issue, and am now going to close it in favour of #6982, which outlines an approach for improving Ghost's config.
